pwiz_tools/Skyline/Model/SkypFile.cs

```diff
@@ -152,20 +155,50 @@ public bool Open(string skypPath, IEnumerable<Server> servers, FormEx parentWind
             }
             catch (Exception e)
             {
-                var message = TextUtil.LineSeparate(Resources.SkypSupport_Open_Failure_opening_skyp_file_, e.Message);
-                MessageDlg.ShowWithException(parentWindow ?? _skyline, message, e);
-                return false;
+                if (skyp != null && e.Message.Contains(AddPanoramaServerMessage(skyp)))
+                {
+                    return AddServerAndOpen(skypPath, e.Message, parentWindow);
+                }
+                else
+                {
+                    var message = TextUtil.LineSeparate(Resources.SkypSupport_Open_Failure_opening_skyp_file_, e.Message);
+                    MessageDlg.ShowWithException(parentWindow ?? _skyline, message, e);
+                    return false;
+                }
             }
         }
 
+        private bool AddServerAndOpen(string skypPath, String message, FormEx parentWindow)
+        {
+            using (var alertDlg = new AlertDlg(message, MessageBoxButtons.OKCancel))
+            {
+                alertDlg.ShowDialog(parentWindow ?? _skyline);
+                if (alertDlg.DialogResult == DialogResult.OK)
+                {
+                    using (var dlg = new ToolOptionsUI(_skyline.Document.Settings))
+                    {
+                        // Let the user add a Panorama server.
+                        dlg.NavigateToTab(ToolOptionsUI.TABS.Panorama);
+                        dlg.ShowDialog(alertDlg);
+                        if (dlg.DialogResult != DialogResult.OK)
+                        {
+                            return false;
+                        }
+                    }
+                    return Open(skypPath, Settings.Default.ServerList /*get the updated server list*/, parentWindow);
+                }
+
+                return false;
+            }
+        }
 
         private void Download(SkypFile skyp, IProgressMonitor progressMonitor, FormEx parentWindow = null)
         {
             var progressStatus =
                 new ProgressStatus(string.Format(Resources.SkypSupport_Download_Downloading__0_, skyp.SkylineDocUri));
             progressMonitor.UpdateProgress(progressStatus);
 
-            if (DownloadClient == null)
+            if (DownloadClient == null || DownloadClient is WebDownloadClient)
```

Comment:
I don't follow this. Why do you create a new DownloadClient when you have one that is a WebDownloadClient. Isn't that the same as always creating one? Seems like a comment here explaining would be helpful.

Reply:
This was for SkypTest to be able to use its custom implementation of IDownloadClient.  I replaced this with a DownloadClientCreator to avoid confusion in this method.